src/ucs/datastruct/ptr_map.inl

```diff
@@ -33,37 +33,233 @@ BEGIN_C_DECLS
 KHASH_IMPL(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*, 1,
            kh_int64_hash_func, kh_int64_hash_equal);
 
-/**
- * Initialize a pointer map.
- *
- * @param [in]  map     Map to initialize.
- * @return      UCS_OK on success otherwise error code as defined by
- *              @ref ucs_status_t.
- */
-static inline ucs_status_t ucs_ptr_map_init(ucs_ptr_map_t *map)
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_create_key(uint64_t *next_id)
+{
+    return (*next_id += UCS_PTR_MAP_KEY_MIN_ALIGN) |
+            UCS_PTR_MAP_KEY_INDIRECT_FLAG;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucs_ptr_hash_put(ucs_ptr_hash_t *hash, ucs_ptr_map_key_t key, void *ptr)
+{
+    khiter_t iter;
+    int ret;
+
+    iter = kh_put(ucs_ptr_map_impl, hash, key, &ret);
+    if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
+        return UCS_ERR_NO_MEMORY;
+    } else if (ucs_unlikely(ret == UCS_KH_PUT_KEY_PRESENT)) {
+        return UCS_ERR_ALREADY_EXISTS;
+    }
+
+    kh_value(hash, iter) = ptr;
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucs_ptr_hash_get(ucs_ptr_hash_t *hash, ucs_ptr_map_key_t key, int extract,
+                 void **ptr_p)
+{
+    khiter_t iter;
+
+    iter = kh_get(ucs_ptr_map_impl, hash, key);
+    if (ucs_unlikely(iter == kh_end(hash))) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    *ptr_p = kh_value(hash, iter);
+    if (extract) {
+        kh_del(ucs_ptr_map_impl, hash, iter);
+    }
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE int
+ucs_ptr_check_direct_put(void *ptr, int direct, ucs_ptr_map_key_t *key)
+{
+    if (ucs_unlikely(!direct)) {
+        return 0;
+    }
+
+    *key = (uintptr_t)ptr;
+    ucs_assert(!(*key & UCS_PTR_MAP_KEY_MIN_ALIGN));
+    ucs_assert(*key != 0);
+    return 1;
+}
+
+static UCS_F_ALWAYS_INLINE int
```

Comment:
maybe return NULL or valid pointer to simplify usability?

Reply:
Removed this method because it doesn't simplify the code.
(It was introduced earlier in this PR, when it was used in two methods. It is used only once now.)